batch2/batch/driver/instance.py

```diff
@@ -105,6 +105,16 @@ def state(self):
         self._state = 'deleted'
         self.instance_pool.adjust_for_add_instance(self)
 
+    async def mark_jobs_ended(self, timestamp):
+        await self.db.execute_update(
+            '''
+UPDATE attempts
+INNER JOIN jobs ON attempts.batch_id = jobs.batch_id AND attempts.job_id = jobs.job_id
+SET start_time = %s, end_time = %s
```

Comment:
This seems wrong: you're setting the start to the timestamp.  It looks like this is called on deactivation/preemption.  The cost of the job is from when it was started until when it was preempted.

Reply:
I can set the start_time to Null. I think the cost will be the same. The trigger left-normalizes any updates and then computes the difference between the new time interval and the old time interval. I could be wrong, but I think this trigger is correct and makes the logic much simpler to deal with. Otherwise, theoretically you could get a preemption event, followed by a mark job complete. I wanted the mark job complete to be handled properly in this case.

Reply:
I'm afraid I'm confused by this function, I don't see a situation where you would want to set the start to the end, or clear the start (forgetting when it started).  I must be confused about how this function is intended to be used.  Based on the name, why isn't it just setting the end only?

This is called on instance preemption/delete.  If a job started and hour ago, and gets preempted now, the user should get charged for an hour of compute.  But clearing start or setting start = end would appear to lose that information.  What am I missing?

Reply:
I can just set the end only. It doesn't affect the ultimate answer (the before update trigger takes care of it), but I agree it's clearer to not set start here at all.